src/transformers/modeling_utils.py

```diff
@@ -2955,6 +2955,19 @@ def from_pretrained(
                 **kwargs,
             )
         else:
+            # In case one passes a config to `from_pretrained` + "attn_implementation"
+            # override the `_attn_implementation` attribute to `attn_implementation` of the kwargs
+            # Please see: https://github.com/huggingface/transformers/issues/28038
+
+            # Overwrite `config._attn_implementation` by the one from the kwargs --> in auto-factory
+            # we pop attn_implementation from the kwargs but this handles the case where users
+            # passes manually the config to `from_pretrained`.
+            config = copy.deepcopy(config)
+
+            if kwargs.get("attn_implementation", None) is not None and getattr(
+                config, "_attn_implementation", None
+            ) != kwargs.get("attn_implementation", None):
```

Comment:
This handles the case where users pass a `config` object to `from_pretrained`. Note `AutoModelxxx.from_pretrained` pops the `attn_impelmentation` from the kwargs in case one do not pass a config, but doesn't if we pass the config. 

Therefore this handles this corner case as well (passing a config --> `attn_implementation` does not get popped + `attn_implementation` through from_pretrained kwargs). If that's the case we should over-write the config's attn_impelmentation by the one from the kwargs assuming the user knows what they are doing. 

https://github.com/huggingface/transformers/blob/main/src/transformers/models/auto/auto_factory.py#L516-L540